### PR TITLE
improvement: resources file url

### PR DESCRIPTION
**Pull request by taufiq**
This PR aims to replace the current way users input the file URL for their resources by introducing a button that they can click and select a list of files from, instead of inputting the URLs themselves.

The PR also:
- [X] Introduces a file modal
image
- [x] Updates the `react-scripts` package to the latest version to allow use of optional chaining
- [X] Introduces `FormFieldFile`, a form field to help users set permalinks of files by selecting files instead of manually keying in the permalink
image
- [X] Validates `FormFieldFile` input
image
- [X] Replace type-specified media components (i.e `ImageCard`/`FileCard`/`FileUploadCard`/`ImageUploadCard`) with the base media components they were constructed from

Combining the `FilesModal` & `ImagesModal` is done in a separate PR (#135)

~This PR should only be worked on after the `revamp-files-tab` PR is approved, as this PR should be built upon the work from the `revamp-files-tab` branch.~

### CODE REVIEW

File: src/layouts/Files.jsx [hunk 1]
```diff
@@ -109,9 +106,10 @@ export default class Files extends Component {
                 />
                 {files.length > 0
                   && files.map((file) => (
-                    <FileCard
+                    <MediaCard
+                      type="file"
+                      media={file}
                       siteName={siteName}
-                      file={file}
                       key={file.fileName}
```
[the remark this answers is not in the document]
Comment (taufiq) [reply src/layouts/Files.jsx]: it's set in the default props of `MediaCard`.

File: src/components/ResourceSettingsModal.jsx [hunk 2]
```diff
@@ -261,15 +276,29 @@ export default class ResourceSettingsModal extends Component {
...
                 <div className="d-flex">
                   <label htmlFor="radio-post" className="flex-fill">
-                    <input type="radio" id="radio-post" name="resource-type" value="post" onClick={this.handlePermalinkFileUrlToggle} checked={!!permalink} />
+                    <input
+                      type="radio"
+                      id="radio-post"
+                      name="resource-type"
+                      value="post"
+                      onClick={this.handlePermalinkFileUrlToggle}
+                      checked={permalink || fileUrl === null}
```
[the remark this answers is not in the document]
Comment (taufiq) [reply src/components/ResourceSettingsModal.jsx]: when there was no logical operators involved, I used double bangs (`!!`) to indicate that the `checked` attribute uses a boolean value. 

Now that there's an OR operator involved (`||`), the expression `permalink || fileUrl === null` would always evaluate to a boolean so the double bangs is not necessary.

File: src/components/ResourceSettingsModal.jsx [hunk 3]
```diff
@@ -93,9 +96,18 @@ export default class ResourceSettingsModal extends Component {
...
+        fileUrl: '',
+      });
     } else {
-      this.setState({ permalink: '/permalink/', fileUrl: null });
+      this.setState({
+        permalink: '/permalink/',
+        fileUrl: null,
+        errors: {
+          fileUrl: '',
+        },
```
[the remark this answers is not in the document]
Comment (taufiq) [reply src/components/ResourceSettingsModal.jsx]: oh yeah but this isn't a new bug.

iirc its already been like that before this PR.

[the remark this answers is not in the document]
Comment (taufiq) [reply src/components/ResourceSettingsModal.jsx]: probably should record that as an issue and fix them all in one PR

File: src/components/FilesModal.jsx [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+      onFileSelect,
+    } = this.props;
+    const { files, selectedFile, stagedFileForUpload } = this.state;
+    return (!!files.length
+      && (
+        <>
+          <div className={elementStyles.overlay}>
+            <div className={mediaStyles.mediaModal}>
+              <div className={elementStyles.modalHeader}>
+                <h1 style={{ flexGrow: 1 }}>Select File</h1>
```
[the remark this answers is not in the document]
Comment (taufiq) [reply src/components/FilesModal.jsx]: It's just a minor tweak to shift the "Select File" button to the right, don't think it's necessary to create a whole CSS class just to do that.

File: src/components/FilesModal.jsx [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+                <LoadingButton
+                  label="Select file"
+                  disabledStyle={elementStyles.disabled}
+                  className={elementStyles.blue}
+                  callback={() => onFileSelect(`/${decodeURIComponent(selectedFile.path)}`)}
+                />
```
[the remark this answers is not in the document]
Comment (taufiq) [reply src/components/FilesModal.jsx]: It's according to Yixin's design

**Comment by taufiq** [comment]
> Refer to comments for requested changes
> 
> One more thing - the `Add a new resource` button does not work because of failed prop types. We made some props required (namely `fileName` and `category`) for the `ResourceSettingsModal` but these aren't present when we click on that `Add a new resource` button. We don't have to fix this in this PR but we should make a note.



> Refer to comments for requested changes
> 
> One more thing - the `Add a new resource` button does not work because of failed prop types. We made some props required (namely `fileName` and `category`) for the `ResourceSettingsModal` but these aren't present when we click on that `Add a new resource` button. We don't have to fix this in this PR but we should make a note.

@kwajiehao I did some digging and it wasn't working not because of the failed prop type, but a bug I introduced in e593255. I've put up a commit that fixes and explains why the `add a new resource` button didn't work ( 16aebfe )

File: src/components/FilesModal.jsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (kwajiehao) [reply src/components/FilesModal.jsx]: Sure, let's make a class for it tho if there's more than one occurrence of this

File: src/components/ResourceSettingsModal.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (kwajiehao) [reply src/components/ResourceSettingsModal.jsx]: Recorded as #136

**Pull request merged**